### ISSUE TITLE
Simple c2s bcast test

### ISSUE THREAD

**Issue by seblamydlc** [RBigData/pbdCS#7]
Hello all,

First of all I would like to congratulate you guys for pbdR. It's a great project and we are using it more and more.

I am running into an issue on a pretty basic script and I figured I would post it here.

I spawn a simple cluster with 2 processes. (mpirun -np 2 Rscript -e "pbdCS::pbdserver()")
The script I am trying to run is the following:
```
my.local.var` <- 'HOHO'
pbdCS::pbdclient()
library(pbdMPI)
init()
c2s(my.local.var)
my.local.var <- bcast(my.local.var)
comm.cat(my.local.var, "\n", all.rank = TRUE)
```

I run into an error at this step:

```
R_zmq_msg_send
 errno: 88 strerror: Socket operation on non-socket
R_zmq_msg_recv errno: 88 strerror: Socket operation on non-socket
Error in barrier() : Message truncated, error stack:
PMPI_Barrier(426)...................: MPI_Barrier(MPI_COMM_WORLD) failed
MPIR_Barrier_impl(308)..............: 
MPIR_Bcast_impl(1369)...............: 
MPIR_Bcast_intra(1199)..............: 
MPIR_Bcast_binomial(149)............: 
MPIC_Recv(109)......................: 
MPIDI_CH3U_Request_unpack_uebuf(605): Message truncated; 4 bytes received but buffer size is 1
Calls: <Anonymous> ... pbd_repl_server -> remoter_repl_server -> evalfun -> barrier -> .Call
```

Am I using it wrong or is this unexpected behaviour?

Thanks!

Sébastien

**Comment by snoweye** [comment]
- It is not recommended to assign values use c2s() unless you really need it! In SPMD, you can just name it. The next is fine with me
```
pbdR> library(pbdMPI)
pbdR> init()
pbdR> a <- "hi"
pbdR> comm.cat(a, all.rank = TRUE)
```
- You need to initial values on all rank as well. `my.local.var` is not known on rank 1 in your case, so you can not `bcast()` it from rank 0. The next is fine with me
```
> a <- "aa"
> pbdCS::pbdclient()

pbdR> library(pbdMPI)
pbdR> init()
pbdR> a <- NULL                                # both ranks have NULL
pbdR> comm.cat(a, all.rank = TRUE)
pbdR> c2s(a)                                      # copy from local to rank 0
pbdR> comm.cat(a, all.rank = TRUE)  # only rank 0 has "aa"
pbdR> a <- bcast(a)
pbdR> comm.cat(a, all.rank = TRUE)  # both ranks have "aa"
```
- Note that I use `pbdZMQ`, `remoter`, and `pbdCS` from my repo at [snoweye](https://github.com/snoweye/) for the above tests.

**Comment by wrathematics** [comment]
Re-opening, because the broken syntax really is reasonable.

We'll have to think about modifying `c2s()` for pbdCS to support this, but it shouldn't be too hard. It may take me a little time to get to, though.

**Comment by snoweye** [comment]
- No. It hangs at `bcast()` ...
- I don't suggest to make `c2s()` to have `bcast()` in it. The same for `s2c()` that should now have a `gather()` in it. Except `pbdCS` has to overwrite `c2s()` and `s2c()` and `pbdMPI` is auto-loaded right after the call to `library(pbdCS)` ...

**Comment by wrathematics** [comment]
Yeah, I mean, it complicates things. But I think it makes sense. At the least, we should have the "collective" `c2s()` and `s2c()` versions with different syntax.

**Comment by seblamydlc** [comment]
Thanks guys,

This snippet wasn't taken from a use case. It occured while I was writing an example on how to use pbdCS. I agree that c2s is probably not a very useful feature in practice.
